docs: improvements on editor public api (#1143)
anatawa12/AvatarOptimizer: .docs/content/docs/developers/make-your-components-compatible-with-aao/index.md

If your tool is a non-destructive tool based on NDMF[^NDMF], you can remove your components before the phases
prior to the Optimizing phase of NDMF or before `com.anatawa12.avatar-optimizer` plugin
(with [`BeforePlugin`][ndmf-BeforePlugin]) in the Optimizing phase.
If your tool removes your components in Optimizing phase, it's highly recommended to specify [`BeforePlugin`][ndmf-BeforePlugin]
even if your default callback order is before `com.anatawa12.avatar-optimizer` plugin.

If your tool is a non-destructive tool not based on NDMF[^NDMF], removing your components before
the NDMF's Optimizing phase is recommended.
In this case, current NDMF executes Optimizing phase in order `-1025`, which is JUST before VRCSDK's `RemoveAvatarEditorOnly`
callback, so your tool should remove components with `IVRCSDKPreprocessAvatarCallback` with `callbackOrder` smaller than `-1025`.

If your components is only for holding data for your edit-mode tool and doesn't affects on build or at runtime,
you can remove your components in `IVRCSDKPreprocessAvatarCallback` as described above, or

> you can simply use [Asset Description] to register your components as safe-to-remove components

Reviewer: ```suggestion
you can simply use [Asset Description] to register your components as safe-to-remove components.
```
ピリオド抜けてる

1. Remove your components before Avatar Optimizer processes as much as possible.

   If your component is not working at runtime, (in other words, it's a build-time or edit mode only component),
   it's mostly better for your tool to process avatar before Avatar Optimizer processes,
   and remove your components before Avatar Optimizer processes.

   Please refer [section below](#remove-component) for more details.

   Avatar Optimizer internally uses this method for most Avatar Optimizer components, 
   that will be processed before Trace and Optimize.

2. Register your components to Avatar Optimizer with API

   If your component is working at runtime, or your tool actually want to process avatar after Avatar Optimizer processes,
   you can register your components to Avatar Optimizer to tell about your components.

   Please refer [section below](#register-component) for more details.

   Avatar Optimizer internally uses this method to keep some components that are processed after Trace and Optimize, 
   and components from Unity, VRCSDK, and other avatar platform components.

3. Register your components as no problems to remove with Asset Description.

>    Since Avatar Optimizer v1.7.0, you can use [Asset Description] to register components only for preserving data

Author: ```suggestion
   Since Avatar Optimizer v1.7.0, you can use [Asset Description] to register components only for holding data
```

   for edit-mode tools, that doesn't effects on build or at runtime.
   If your tool process nothing at build time or runtime, you can use this to register your components instead of
   removing your components before Avatar Optimizer processes.
   Please refer [Asset Description] for more details.

   If your tool process something at build time, registering with Asset Description is not recommended.
   Using Asset Description for components that process something at build time may unexpectedly
   remove your components and disables your tool if the execution order is incorrect or unexpectedly changed.

   This method is internally used by Avatar Optimizer to keep compatibility with well-known edit-mode tools.

[Asset Description]: ../asset-description

### Removing your components {#remove-component}

You can remove your components with [`DestroyImmediate`][DestroyImmediate] to remove your components.

> There is several ways to process and remove your component from avatar before Avatar Optimizer processes on build.

Author: ```suggestion
There are several ways to process and remove your components from avatar before Avatar Optimizer processes on build. You can use [`DestroyImmediate`][DestroyImmediate] method for removing your components.
```

If your tool is a non-destructive tool based on NDMF[^NDMF], you can remove your components before the phases
prior to the Optimizing phase of NDMF or before `com.anatawa12.avatar-optimizer` plugin
(with [`BeforePlugin`][ndmf-BeforePlugin]) in the Optimizing phase.
If your tool removes your components in Optimizing phase, it's highly recommended to specify [`BeforePlugin`][ndmf-BeforePlugin]
even if your default callback order is before `com.anatawa12.avatar-optimizer` plugin.

If your tool is a non-destructive tool not based on NDMF[^NDMF], removing your components before
the NDMF's Optimizing phase is recommended.
In this case, current NDMF executes Optimizing phase in order `-1025`, which is JUST before VRCSDK's `RemoveAvatarEditorOnly`

> callback, so your tool should remove components with `IVRCSDKPreprocessAvatarCallback` with `callbackOrder` smaller than `-1025`.

Author: https://github.com/anatawa12/AvatarOptimizer/pull/1143/files#r1731004497

> If your components is only for holding information for your edit mode tool and has no meaning on the build time,

Author: ```suggestion
If your components is only for holding data for your edit-mode tool and doesn't affects on build or at runtime,
```

you can remove your components in `IVRCSDKPreprocessAvatarCallback` as described above, or

> you can somply use [Asset Description] to register your components to be removed.

Author: as no problems違和感があってこっちにしてた。

```suggestion
you can simply use [Asset Description] to register your components as safe-to-remove components
```

とか？

> you can somply use [Asset Description] to register your components to be removed.

Reviewer: よさそう
前方もそれかな。

> you can somply use [Asset Description] to register your components to be removed.

Author: 前方ってどこです？

> you can somply use [Asset Description] to register your components to be removed.

Reviewer: `3. Register your components as no problems to remove with Asset Description.`
辺りですかね

> you can somply use [Asset Description] to register your components to be removed.

Author: 事前説明がないところだとsafe-to-removeで伝わるか怪しい気がする(文脈があれば大丈夫ってくらいの表現だと思ってる)のでそこは変えちゃうと分かりづらいと思う

> you can somply use [Asset Description] to register your components to be removed.

Reviewer: ふーむ
